patches/minecraft/net/minecraft/world/level/levelgen/feature/treedecorators/AlterGroundDecorator.java.patch

```diff
@@ -0,0 +1,15 @@
+--- a/net/minecraft/world/level/levelgen/feature/treedecorators/AlterGroundDecorator.java
++++ b/net/minecraft/world/level/levelgen/feature/treedecorators/AlterGroundDecorator.java
+@@ -71,8 +_,10 @@
+    private void m_225973_(TreeDecorator.Context p_225974_, BlockPos p_225975_) {
+       for(int i = 2; i >= -3; --i) {
+          BlockPos blockpos = p_225975_.m_6630_(i);
+-         if (Feature.m_65788_(p_225974_.m_226058_(), blockpos)) {
+-            p_225974_.m_226061_(blockpos, this.f_69303_.m_213972_(p_225974_.m_226067_(), p_225975_));
++         net.minecraft.world.level.block.state.BlockState blockState = this.f_69303_.m_213972_(p_225974_.m_226067_(), p_225975_);
++         net.minecraftforge.event.level.AlterGroundEvent event = net.minecraftforge.event.ForgeEventFactory.alterGround(p_225974_.m_226058_(), p_225974_.m_226067_(), p_225975_, blockState);
++         if (!event.isCanceled() && Feature.m_65788_(p_225974_.m_226058_(), blockpos)) {
++            p_225974_.m_226061_(blockpos, blockState);
```

Comment:
In general, I think the event should go in when the dirt block is found, meaning it'll be in the `dirt` tag, but if they want it to become a block they can subscribe to the event. Otherwise, this'll get called even on the tree blocks.